R-packages/evalcast/R/download_signals.R

```diff
@@ -1,30 +1,90 @@
-#' Wrapper around `covidcast::covidcast_signal()` with additional logging.
+#' Wrapper around `covidcast::covidcast_signal()` with additional logging and caching.
 #' @param ... arguments to be passed to `covidcast::covidcast_signal()`.
+#' @param offline_signal_dir the directory that stores the cached data for each
+#' (signal, forecast day) pair. If this is null, no caching is done and the data is
+#' downloaded from covidcast.
 #' @return `covidcast_signal` data frame.
 #'
-download_signal <- function(...) {
+#' @importFrom fs dir_create
+#' @importFrom rlang warn inform
+#'
+#' @export
+download_signal <- function(..., offline_signal_dir=NULL) {
   args <- list(...)
-  if (is.null(args$start_day)) {
-    msg <- stringr::str_glue(
-      "Downloading {signal} from covidcast through {end}.",
-      start = args$start_day,
-      end = args$end_day,
-      signal = args$signal,
-      .sep = " "
-    )
-  }
-  else {
-    msg <- stringr::str_glue(
-      "Downloading {signal} from covidcast for period from",
-      "{start} to {end}.",
-      start = args$start_day,
-      end = args$end_day,
-      signal = args$signal,
-      .sep = " "
+  # Mirror covidcast_signal defaults
+  if (is.null(args$as_of)) args$as_of <- Sys.Date()
+  if (is.null(args$geo_type)) args$geo_type <- "county"
+  if (!is.null(offline_signal_dir) & is.null(args$end_day)) stop("end_day must be set.")
+
+  if (is.null(offline_signal_dir)) {
+    if (is.null(args$start_day)) {
+      msg <- stringr::str_glue(
+        "Downloading {signal} from covidcast through {end}.",
+        end = args$end_day,
+        signal = args$signal,
+        .sep = " "
+      )
+    }
+    else {
+      msg <- stringr::str_glue(
+        "Downloading {signal} from covidcast for period from",
+        "{start} to {end}.",
+        start = args$start_day,
+        end = args$end_day,
+        signal = args$signal,
+        .sep = " "
+      )
+    }
+
+    message(msg)
+    df <- suppressMessages({covidcast_signal_wrapper(...)})
+  } else {
+    signal_fpath <- file.path(offline_signal_dir, args$geo_type, sprintf("%s_%s_%s_%s.RDS", args$data_source, args$signal, args$end_day, args$as_of))
```

Comment:
To the first comment, not sure we can do much about guarding a user using positional arguments aside from explicitly writing the arguments instead of using the ellipsis.

To the second, you're right, it's a no-op at this point. I like the idea of having the file tagged with the as_of date instead and then the end_day can actually filter values if end_day < as_of.

Reply:
Regarding the first: we move from having a bunch of arguments hidden in the dots to explicitly having them as parameters, or check for unnamed arguments and throw a more descriptive error.

Regarding the second: I like the idea of tagging with as_of date only.  But I think the post-cache-fetch filter would need to be something like .data$time_value < .env$end_day, not involving as_of at all.

Reply:
I addressed these issues in a recent update.